### ISSUE TITLE
sql: TestTxnObeysLeaseExpiration is disabled

### ISSUE THREAD

**Issue by tschottdorf** [cockroachdb/cockroach#7031]
The following test appears to have failed:

[#18584](https://circleci.com/gh/cockroachdb/cockroach/18584):

```
I160603 16:45:49.136637 storage/intent_resolver.go:104  [node=1,store=1,range=7] on DeleteRange: failed to push "sql/executor.go:422 sql txn" id=91996cd5 key=/Table/51/1/"c"/0 rw=true pri=0.03708904 iso=SNAPSHOT stat=PENDING epo=0 ts=1464972350.128315874,540 orig=1464972348.926062474,0 max=1464972348.926062474,0 wto=false rop=false
I160603 16:45:49.136827 kv/dist_sender.go:983  application error: txn "sql/executor.go:422 sql txn" id=ca317e96 key=/Table/51 rw=false pri=0.01976239 iso=SNAPSHOT stat=PENDING epo=0 ts=1464972350.128315874,631 orig=1464972350.128315874,631 max=1464972350.128315874,631 wto=false rop=false failed to push "sql/executor.go:422 sql txn" id=91996cd5 key=/Table/51/1/"c"/0 rw=true pri=0.03708904 iso=SNAPSHOT stat=PENDING epo=0 ts=1464972350.128315874,540 orig=1464972348.926062474,0 max=1464972348.926062474,0 wto=false rop=false
W160603 16:45:49.136930 kv/dist_sender.go:728  failed to invoke BeginTransaction [/Table/51,/Min), DeleteRange [/Table/51,/Table/52): txn "sql/executor.go:422 sql txn" id=ca317e96 key=/Table/51 rw=false pri=0.01976239 iso=SNAPSHOT stat=PENDING epo=0 ts=1464972350.128315874,631 orig=1464972350.128315874,631 max=1464972350.128315874,631 wto=false rop=false failed to push "sql/executor.go:422 sql txn" id=91996cd5 key=/Table/51/1/"c"/0 rw=true pri=0.03708904 iso=SNAPSHOT stat=PENDING epo=0 ts=1464972350.128315874,540 orig=1464972348.926062474,0 max=1464972348.926062474,0 wto=false rop=false
I160603 16:45:49.137093 client/db.go:474  failed batch: txn "sql/executor.go:422 sql txn" id=ca317e96 key=/Table/51 rw=false pri=0.03708899 iso=SNAPSHOT stat=PENDING epo=1 ts=1464972350.128315874,631 orig=1464972350.128315874,631 max=1464972352.128315874,631 wto=false rop=false failed to push "sql/executor.go:422 sql txn" id=91996cd5 key=/Table/51/1/"c"/0 rw=true pri=0.03708904 iso=SNAPSHOT stat=PENDING epo=0 ts=1464972350.128315874,540 orig=1464972348.926062474,0 max=1464972348.926062474,0 wto=false rop=false
I160603 16:45:49.137558 storage/engine/rocksdb.go:178  closing in-memory rocksdb instance
--- FAIL: TestTxnObeysLeaseExpiration (0.43s)
    lease_test.go:660: pq: restart transaction: txn "sql/executor.go:422 sql txn" id=ca317e96 key=/Table/51 rw=false pri=0.03708899 iso=SNAPSHOT stat=PENDING epo=1 ts=1464972350.128315874,631 orig=1464972350.128315874,631 max=1464972352.128315874,631 wto=false rop=false failed to push "sql/executor.go:422 sql txn" id=91996cd5 key=/Table/51/1/"c"/0 rw=true pri=0.03708904 iso=SNAPSHOT stat=PENDING epo=0 ts=1464972350.128315874,540 orig=1464972348.926062474,0 max=1464972348.926062474,0 wto=false rop=false
=== RUN   TestLogic
Running logic test on file: testdata/aggregate
I160603 16:45:49.142079 storage/engine/rocksdb.go:137  opening in memory rocksdb instance
W160603 16:45:49.145176 gossip/gossip.go:897  not connected to cluster; use --join to specify a connected node
I160603 16:45:49.146143 server/node.go:401  store store=0:0 ([]=) not bootstrapped
W160603 16:45:49.146210 gossip/gossip.go:897  not connected to cluster; use --join to specify a connected node
I160603 16:45:49.147663 storage/replica_command.go:1515  range 1: new leader lease replica {1 1 1} 1970-01-01 00:00:00 +0000 UTC 406936h45m50.147368359s following replica {0 0 0} 1970-01-01 00:00:00 +0000 UTC 0 [physicalTime=2016-06-03 16:45:49.147598788 +0000 UTC]
I160603 16:45:49.148256 server/node.go:327  **** cluster {e1fd50c3-660d-444a-bb3e-c34595dfd7d5} has been created
I160603 16:45:49.148278 server/node.go:328  **** add additional nodes by specifying --join=127.0.0.1:60351
--
I160603 16:46:33.058688 kv/dist_sender.go:983  application error: retry txn "sql/executor.go:422 sql txn" id=366cfaa8 key=/Table/51/1/"g"/0 rw=true pri=0.04010373 iso=SERIALIZABLE stat=PENDING epo=0 ts=1464972393.041274183,0 orig=1464972393.041274183,0 max=1464972393.041274183,0 wto=false rop=false
W160603 16:46:33.058781 kv/dist_sender.go:728  failed to invoke EndTransaction [/Table/51/1/"g"/0,/Min): retry txn "sql/executor.go:422 sql txn" id=366cfaa8 key=/Table/51/1/"g"/0 rw=true pri=0.04010373 iso=SERIALIZABLE stat=PENDING epo=0 ts=1464972393.041274183,0 orig=1464972393.041274183,0 max=1464972393.041274183,0 wto=false rop=false
I160603 16:46:33.058926 client/db.go:474  failed batch: retry txn "sql/executor.go:422 sql txn" id=366cfaa8 key=/Table/51/1/"g"/0 rw=true pri=0.04010373 iso=SERIALIZABLE stat=PENDING epo=1 ts=1464972393.041274183,0 orig=1464972393.041274183,0 max=1464972393.091274183,0 wto=false rop=false
I160603 16:46:33.060950 storage/engine/rocksdb.go:178  closing in-memory rocksdb instance
--- PASS: TestRollbackInRestartWait (0.25s)
FAIL
FAIL    github.com/cockroachdb/cockroach/sql    53.480s
I160603 16:45:56.578970 rand.go:76  Random seed: 999005719796818881
=== RUN   TestProcessExpression
--- PASS: TestProcessExpression (0.01s)
=== RUN   TestServer
I160603 16:45:56.588761 storage/engine/rocksdb.go:137  opening in memory rocksdb instance
W160603 16:45:56.593957 gossip/gossip.go:897  not connected to cluster; use --join to specify a connected node
I160603 16:45:56.598691 server/node.go:401  store store=0:0 ([]=) not bootstrapped
W160603 16:45:56.603251 gossip/gossip.go:897  not connected to cluster; use --join to specify a connected node
I160603 16:45:56.604011 storage/replica_command.go:1515  range 1: new leader lease replica {1 1 1} 1970-01-01 00:00:00 +0000 UTC 406936h45m57.603322794s following replica {0 0 0} 1970-01-01 00:00:00 +0000 UTC 0 [physicalTime=2016-06-03 16:45:56.603845469 +0000 UTC]
I160603 16:45:56.605011 server/node.go:327  **** cluster {bfa09169-a0c3-4e38-8f60-7bfc722230a4} has been created

```

Please assign, take a look and update the issue accordingly.

**Comment by andreimatei** [comment]
@vivekmenezes is this you? :)

**Comment by vivekmenezes** [comment]
I can take a look at it

**Comment by vivekmenezes** [comment]
The test runs two SNAPSHOT transactions sequentially (not in parallel), with each transaction being interrupted by another transaction that runs a read and pushes each transaction forward so that they both eventually are unable to commit because of a deadline exceeded error.

1. INSERT INTO t.kv VALUES ('c', 'd')

this transaction ends with a "pq: restart transaction: txn aborted" error as is expected because its transaction deadline expires. Before its deadline expires, it is pushed by another transaction that runs a read. All as expected.

2. TRUNCATE TABLE t.kv

This second statement fails right at the start (not a deadline exceeded error) because it is unable to push the first transaction. But the first transaction should have aborted not requiring the push in the first place.

@tschottdorf is it possible for aborts to not be recorded. I'm assuming yes.

**Comment by tschottdorf** [comment]
Should not, but in fact currently is happening on deadline exceeded. Relevant excerpt from `replica_command.go`:

```go
	if isEndTransactionExceedingDeadline(reply.Txn.Timestamp, args) {
		reply.Txn.Status = roachpb.ABORTED
		// FIXME(#3037):
		// If the deadline has lapsed, return all the intents for
		// resolution. Unfortunately, since we're (a) returning an error,
		// and (b) not able to write on error (see #1989), we can't write
		// ABORTED into the master transaction record, which remains
		// PENDING, and that's pretty bad.
		return reply, roachpb.AsIntents(args.IntentSpans, reply.Txn), roachpb.NewTransactionAbortedError()
	}
```

**Comment by bdarnell** [comment]
That comment references #3037 and #1989, both of which are closed. Is there an open issue for this or should we reopen #3037? 

Why does this pass most of the time and only fail non-deterministically? Isn't this test supposed to hit the deadline every time?

**Comment by vivekmenezes** [comment]
I see no reason to reopen any of the issues.

I see why it's happening, but it isn't clear why this is happening non-deterministically. I'll take a look

**Comment by vivekmenezes** [comment]
```
W160615 16:40:28.674891 storage/intent_resolver.go:319  [node=1,store=1,range=6] failed to resolve intents: key range /Table/51/1/"a"/2/1-/Min outside of bounds of range /Table/50-/Table/51
```

when a transaction fails on a commit() due to a deadline exceeded error, it attempts to cleanup its intents. Occasionally the intent cleanup fails leading to intents that interfere and push future transactions. I think this behavior is okay and I suspect we wont find  a simple solution to this problem anytime soon. I'm going to just restructure this test to get around the problem.

**Comment by andreimatei** [comment]
Is this failure to push expected? Looks like something split table 51(the first user table I think) from the system table?

**Comment by vivekmenezes** [comment]
This is a failure to resolve intents, and I think it's because the key being resolved is no longer on this range following a split. I think resolving intents is a best effort operation and so we can expect unresolved intents every once in a while.

**Comment by andreimatei** [comment]
Sure, but I'm just curious what's really going on, mostly for my own edification.
If table 51 got on the right of an async split, what's preventing the same split for causing the test's txn to fail when doing some other operation?

**Comment by vivekmenezes** [comment]
I cant answer that question. Im not familiar with the range  splitting
system.

**Comment by bdarnell** [comment]
These errors are fairly benign - we just log a warning and continue; the next transaction to hit those intents will clean them up and retry. We see the RangeKeyMismatch at this level because the intent resolver bypasses DistSender for intents it believes to be local. However, now that we have the local server optimization to bypass the RPC layer, it may not be worthwhile to special-case local intents. We can just send all intents through the DistSender, which knows how to handle races with splits.

**Comment by andreimatei** [comment]
So future txns running into these intents - are they guaranteed to be able to resolve them? They'll likely fail to resolve them in exactly the same way (bypass the DistSender and get the `RangeKeyMismatch` err) right?

**Comment by bdarnell** [comment]
The split will have completed (or aborted) by then, so they'll be able to make the correct decision about whether the intent is local. This is just a race when async intent resolution happens concurrently with a split.

**Comment by andreimatei** [comment]
Hmmm but then I'm not sure why the test is failing with that retryable error. Maybe it's unrelated to this failure to push...

**Comment by bdarnell** [comment]
If you're getting "failed to push" errors, that means the first transaction is still pending. It's unrelated to the split. The cause is in @tschottdorf's first comment: when we exceed the deadline, we return a TransactionAbortedError, but don't persist the aborted state to disk so no one else can resolve our intents until the transaction times out.

The question above was why this is non-deterministic; the answer is that pushes are non-deterministic due to the random priority. 

I think the ultimate fix is to do the cleanup contemplated in #3037, but we may be able to make a quick fix by adding a flag to this TransactionAbortedError to tell TxnCoordSender to send a real EndTransaction. Or even quicker, return a non-retryable error when transaction deadlines are exceeded. I've had my doubts about retries here because there's a good chance that the transaction is just too large to complete within its lease.

**Comment by vivekmenezes** [comment]
The reason for the intent cleanup is still a mystery because the transaction that hits a commit failure gets aborted through a short-circuit that should not be clearing up intents. I'll investigate this further.

**Comment by vivekmenezes** [comment]
I'm pretty sure an attempt is being made to resolve intents through this code at

https://github.com/cockroachdb/cockroach/blob/master/storage/replica.go#L2046

```
	// On the replica on which this command originated, resolve skipped intents
	// asynchronously - even on failure.
	if originReplica.StoreID == r.store.StoreID() {
		r.store.intentResolver.processIntentsAsync(r, intents)
	}
```

If I remove this code the test fails immediately.

This is the reason for not seeing the intents unless the intent resolution itself fails, which is when the intents remain and the test fails.

Do we think the intended behavior is okay?

**Comment by vivekmenezes** [comment]
I suppose we should fix this by cleaning up the intents when the transaction aborts, but then again I'm not expecting deadlines to be exceeded often, and on top of that intent cleanup failures to occur. The options are:
1. fix the underlying problem (low priority?)
2. fix the test through a hack because we want to keep the test running

**Comment by bdarnell** [comment]
I think we care more about the underlying problem than the test alone. Transactions exceeding their deadline are pretty rare, and a hacky test just gives us a false sense of security if the real-world behavior of this scenario is bad.

I think the best quick fix here (as long as it works) is to return a non-retryable error when the deadline is exceeded. This will cause the higher level systems to send a real abort and not leave the transaction record in the pending state. That should fix this test as well as keeping transactions from getting stuck in any real-world encounters of this issue. It will prevent automatic retries, but since we have a check on the minimum lease time remaining I think that's OK (it will only affect long-running transaction and there's no guarantee that the retry will be fast enough to succeed).

**Comment by tamird** [comment]
@vivekmenezes this test is still skipped. Could you please look into restoring it?

**Comment by tamird** [comment]
Unable to reproduce this in GCE with `make stress PKG=./storage TESTS=TestRaftAfterRemoveRange STRESSFLAGS='-p 100 -stderr -maxfails 1'` after 20k runs.
